spring-social-zotero/src/main/java/org/springframework/social/zotero/api/impl/GroupsTemplate.java

```diff
@@ -350,4 +412,42 @@ public void deleteItem(String groupId, String citationKey, Long citationVersion)
         }
 
     }
+    
+    private ZoteroUpdateItemsStatuses getStatusesFromResponse(List<ItemCreationResponse> responses,
+            List<String> itemsKeys) {
+        // Zotero is not sending failed items keys in response. So, we have to compute
+        // failed items keys on our own using success and unchanged keys
+
+        ZoteroUpdateItemsStatuses statuses = new ZoteroUpdateItemsStatuses();
+        List<String> successKeys = new ArrayList<>();
+        List<String> unchangedKeys = new ArrayList<>();
+        List<String> failedMessages = new ArrayList<String>();
+        Set<String> failedCodes = new HashSet<String>();
+        
+        for (ItemCreationResponse response : responses) {
+            successKeys.addAll(extractItemKeys(response.getSuccess(), e -> e.getValue()));
+            unchangedKeys.addAll(extractItemKeys(response.getUnchanged(), e -> e.getValue()));
+            failedMessages.addAll(response.getFailed().entrySet().stream().map(e -> e.getValue().getMessage())
+                    .collect(Collectors.toList()));
+            failedCodes.addAll(response.getFailed().entrySet().stream().map(e -> e.getValue().getCode().toString())
+                    .collect(Collectors.toSet()));
```

Comment:
I believe that keeping codes and messages separate could become an issue since I don't think that `entrySet` guarantees that you get the same order of elements every time. Additionally, using `toSet`, a set also doesn't guarantee an order. This means that you might end up with a list of messages and a list of codes which are in different orders (so the first code might not be the code for the first message). I would keep them together, so that whoever uses the returned values can decide what they want to do with codes and/or messages.

Reply:
Initially I was thinking of sending a FailedResponse in the below shape, but Zotero hasn't been sending failed item keys. It's really hard to tie back failed item key to its corresponding code and message. I could still tie code and message together and send the response back but in most cases the code and message are just duplicated as most of them fail for same reason.

```
class FailedResponse {
private String itemKey;
private String code;
private String message;
}
```

Here is the raw response from Zotero in the case of failed scenarios.

```
{
    "successful": {},
    "success": {},
    "unchanged": {},
    "failed": {
        "0": {
            "code": 404,
            "message": "Item doesn't exist (expected version 2602; use 0 instead)"
        },
        "1": {
            "code": 412,
            "message": "Item has been modified since specified version (expected 2602, found 2763)"
        }
    }
}
```